File: Challenge-2022-03/01-counting-bits.cpp
#include "../header.h"

class Solution {
public:
    vector<int> countBits(int n) {
        if (n == 0) {
            return {0};
        }
        vector<int> res(n+1, 0);
        for (int i = 1; i <= n; ++i) {
            if ((i & 1) == 0) {
                res[i] = res[i >> 1];
            } else {
                res[i] = 1 + res[i >> 1];
            }
        }
        return res;
    }
};

File: Challenge-2023-09/01-counting-bits.go
package main

func countBits(n int) []int {
	res := make([]int, n+1)
	res[0] = 0
	for i := 1; i <= n; i++ {
		if i%2 == 0 {
			res[i] = res[i/2]
		} else {
			res[i] = res[i/2] + 1
		}
	}
	return res
}
